docs: editorial love for IBC overview

barriebyron opened the pull request:
Apply technical style standards, present tense, fix typos, and use consistent IBC language

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

We all know that IBC is how blockchains speak to each other. The world is watching, let's use our best editorial content quality standards for this essential content. 

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
+
+- Bind to a port or ports
+- Define your own packet data
+- Define optional acknowledgement structs
```
I wasn't sure what is optional, the acknowledgement structs or defining them

AdityaSripal replied:
You can define your own custom acknowledgement struct if you wish, or you can use the default one provided by core IBC.

barriebyron replied:
```suggestion
- Use the default acknowledgment struct provided by core IBC, or optionally define your own custom acknowledgment struct
```

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
 The connection handshake is responsible for verifying that the light clients on each chain are
 correct for their respective counterparties. Connections, once established, are responsible for
-facilitation all cross-chain verification of IBC state. A connection may be associated with any
+facilitation all cross-chain verification of IBC state. A connection can be associated with any
```
I changed all instances of may to can
per Google developer documentation style guide advice for using can https://developers.google.com/style/word-list#letter-c 
and  may https://developers.google.com/style/word-list#may

key point

> may
> In general, reserve _may_ for official policy or legal considerations. To convey possibility, use _can_ or _might_ instead. To convey permission, use _can_ instead. See also can, might, and should.

damiannolan replied:
Should this not read as: "Connections, once established, are responsible for **facilitating** all cross-chain verification of IBC state"?

Another small point which I am unsure of but I assumed previously - I thought the use of `may` here was as per [RFC-2119](https://datatracker.ietf.org/doc/html/rfc2119). @AdityaSripal maybe can you confirm?

AdityaSripal replied:
> Connections, once established, are responsible for facilitating all cross-chain verification of IBC state

Yea that's right.

`MAY` implied optional as mentioned in the RFC, but i don't believe we stuck to a very strict technical writing standard so I defer to experts on the best way to convey this.

barriebyron replied:
I'll use the gerund verb facilitating as recommended, good catch @damiannolan but then we need plural "verifications"

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
-must use for committing IBC messages is defined in
-[ICS-24](https://github.com/cosmos/ics/tree/master/spec/core/ics-024-host-requirements) and the proof
-format that all implementations must be able to produce and verify is defined in this [ICS-23 implementation](https://github.com/confio/ics23).
+communicate, a blockchain commits some state to a specifically defined path that is reserved for a
```
I applied the guideline to use present tense, per https://developers.google.com/style/tense

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
-
-While the above is useful background information, IBC modules do not need to interact at all with
+other. Thus, IBC must authenticate module actions on ports and channels so that only modules with the
+appropriate permissions can use them. This module action authentication is accomplished using a [dynamic
```
for "This" I think we mean "This module action authentication"

beware the naked "This" especially for complex sentences

> this, that
> Where possible, put a noun after this or that for clarity. If doing so results in clunky prose, then don't do it; but even then, try thinking about what the noun would be. If you aren't sure what noun this or that refers to, then consider rephrasing.

see https://developers.google.com/style/word-list#letter-t

AdityaSripal replied:
```suggestion
appropriate permissions can use them. This module authentication is accomplished using a [dynamic
```

Seems like `module authentication` is more appropriate

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
+appropriate permissions can use them. This module action authentication is accomplished using a [dynamic
+capability store](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-003-dynamic-capability-store.md). Upon binding to a port or
+creating a channel for a module, IBC returns a dynamic capability that the module must claim in
+order to use that port or channel. This dynamic capability store prevents other modules from using that port or channel since
```
another unclear reference "This"
I guessed that we are talking about "This dynamic capability store" 

But please teach me if I'm off track

AdityaSripal replied:
```suggestion
order to use that port or channel. The dynamic capability module prevents other modules from using that port or channel since
```

barriebyron reviewed (commented):
some explanation for my changes, teach me where I'm off track, I aim to help IBC content be the best content

crodriguezvega commented:
Thanks @barriebyron for giving some ❤️ to our documentation!

AdityaSripal commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
+
+This brief explanation of the lower levels of the
+stack is provided to give application developers a high-level understanding of the IBC
+protocol. Details on the abstraction layer are most relevant for application
+developers (channels and ports) and describe how to define custom packets and
```
```suggestion
protocol. The channel and port abstraction layer is most relevant to application developers, here we will define custom packets and
```

The second sentence didn't make sense to me.

AdityaSripal commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
+- Bind to a port or ports
+- Define your own packet data
+- Define optional acknowledgement structs
+- Know how to encode and decode the packet data
```
```suggestion
- Standardize an encoding of the packet data
```

AdityaSripal replied:
or you could say `Define` to keep it consistent

AdityaSripal replied:
Though this point seems like it should be rolled into `Defining packet data struct`.

Perhaps `Define packet data struct and encoding`?

The same is true for acknowledgment

AdityaSripal commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
-Modules may choose which channels they wish to communicate over with, thus IBC expects modules to
-implement callbacks that are called during the channel handshake. These callbacks may do custom
-channel initialization logic, if any return an error, the channel handshake will fail. Thus, by
+Modules can choose which channels they wish to communicate over with, thus IBC expects modules to
```
```suggestion
Modules can choose which channels they wish to communicate over, thus IBC expects modules to
```

AdityaSripal reviewed (changes requested):
Thank you for bringing this doc more in line with technical writing standards! I have a few comments here that we should resolve before merging. If you disagree with any suggestion, please comment on them and we can resolve it there.

Thanks again!

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -10,131 +10,147 @@ Learn what IBC is, its components and use cases. {synopsis}
+- Define your own packet data
+- Define optional acknowledgement structs
+- Know how to encode and decode the packet data
+- Implement the `IBCModule` interface
```
@crodriguezvega in this case, we can apply the Google developer documentation technical content style standards and add a period (.) at the end of each list item since the items include a verb.

My preference is to omit punctuation at the end of sentence fragments in lists; the key point is to be consistent throughout the content set. 

https://developers.google.com/style/periods?hl=en#bulleted-lists

barriebyron commented:
hi @AdityaSripal and @crodriguezvega thanks for the feedback and attention.
IBC docs are so essential to building the internet of blockchains... 

I have added a few more in-line questions and made these changes:

- used consistent spelling for acknowledgments 
- fixed some misplaced commas
- tightened up some loose sentences
- applied suggestion to add punctuation to end of list items
- typo fixes that I didn't catch on first round

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,186 @@ parent:
+The acknowledgment can encode whether the packet processing succeeded or failed, along with additional information that allows the sender module to take appropriate action.
 
-Once the acknowledgement has been written by the receiving chain, a relayer will relay the acknowledgement back to the original sender module which will then execute application-specific acknowledgment logic using the contents of the acknowledgement. This may involve rolling back packet-send changes in the case of a failed acknowledgement (refunding senders).
+After the acknowledgment has been written by the receiving chain, a relayer relays the acknowledgment back to the original sender module which  then executes application-specific acknowledgment logic using the contents of the acknowledgment. This can involve rolling back packet-send changes in the case of a failed acknowledgment (refunding senders).
```
The unclear reference (the notorious naked "This") here makes me wonder what the subject is (I couldn't even guess... "This logic"? or "This relay"?)

please r/w or provide the subject of "This" in  

"This can involve rolling"

crodriguezvega replied:
I think it refers to the execution of the acknowledgment logic. So maybe something like `This acknowledgment logic can involve rolling...`.

There is also an extra space in `which  then`.

AdityaSripal replied:
Maybe we should replace that sentence with:

`Application developers can roll back packet-send changes in the case of a failed acknowledgement (e.g. refunding senders in ICS20).`

barriebyron replied:
yes better, the more I read this content, the move love it needs!

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,186 @@ parent:
+Modules can also choose to write application-specific acknowledgments upon processing a packet. This acknowledgment can be done synchronously on `OnRecvPacket` if the module processes packets as soon as they are received from IBC module. Or acknowledgments can be done asynchronously if module processes packets at some later point after receiving the packet.
 
-Regardless, this acknowledgement data is opaque to IBC much like the packet `Data` and will be treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgemnet in such a way that the sender module can decode it correctly. This should be decided through version negotiation during the channel handshake.
+Regardless, this acknowledgment data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgment so that the sender module can decode it correctly. This should be decided through version negotiation during the channel handshake.
```
```suggestion
Regardless, this acknowledgment data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgment so that the sender module can decode it correctly. Define this encoding through version negotiation during the channel handshake.
```

barriebyron replied:
I probably guessed incorrectly but again the unclear reference "This" in "This should be decided"

are we talking about the encoding? or ? and how is it decided?

crodriguezvega replied:
I think we're talking about how it is decided the encoding, so that both parties know how to encode/decode to understand each other.

AdityaSripal replied:
`The encoding should be negotiated between the two parties during version negotiation in the channel handshake.

barriebyron replied:
since this is the overview topic, I suggest that we are too specific here.

Readers can learn more at https://docs.cosmos.network/master/ibc/custom.html#channel-handshake-version-negotiation 
---
oh "should" is such an ambiguous word, let's avoid per this advice
https://developers.google.com/style/word-list?hl=en#letter-s 

**should, should be
Generally avoid.**
![image](https://user-images.githubusercontent.com/10067215/142517735-4665c117-9deb-46ae-a413-2aede50b63ea.png)

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,186 @@ parent:
+then no packet at sequence `k > n` can be successfully received without violating the contract of ORDERED channels that packets are processed in the order that they are sent. Since ORDERED channels enforce this invariant, a proof that sequence `n` hasn't been received on the destination chain by the specified timeout of packet `n` is sufficient to timeout packet `n` and close the channel.
 
-In the UNORDERED case, packets may be received in any order. Thus, IBC will write a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information, it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, one must provide a proof that a packet  ...
+- In UNORDERED channels, packets can be received in any order. Thus, IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information; it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, one must provide a proof that a packet re ...
```
I didn't tackle this grammatical challenge, but now that we are all focused on clarity and overall improvements, let's see if we can't improve here 

```suggestion
- In UNORDERED channels, packets can be received in any order. Thus, IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information; it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, proof is required that a packet receipt does not exist for the packet's sequence by the specified timeout. Of course, timing out a packet on an UNORDERED channel simply triggers the application-specific timeout logic for that packet and does not close the channel.
```

barriebyron reviewed (commented):
take it from here, ya'll!

Thanks for the editorial partnering for awesomeness

AdityaSripal commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,186 @@ parent:
+then no packet at sequence `k > n` can be successfully received without violating the contract of ORDERED channels that packets are processed in the order that they are sent. Since ORDERED channels enforce this invariant, a proof that sequence `n` hasn't been received on the destination chain by the specified timeout of packet `n` is sufficient to timeout packet `n` and close the channel.
 
-In the UNORDERED case, packets may be received in any order. Thus, IBC will write a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information, it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, one must provide a proof that a packet  ...
+- In UNORDERED channels, packets can be received in any order. Thus, IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information; it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, one must provide a proof that a packet re ...
```
```suggestion
- In UNORDERED channels, packets can be received in any order. Thus, IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information; it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. In order to timeout a packet on an UNORDERED channel, a proof is required that a packet receipt **does not exist** for the packet's sequence by the specified timeout deadline. Timing out a packet on an UNORDERED channel prevents any further receiving of the packet, triggers the application-specific timeout logic for that packet, but does not close the channel.
```

barriebyron replied:
see how the rewrite is now... let's try to write concisely what is happening. Instead of "deadline" let's use "timeout" (since it's timeout height, not a date-based deadline)

let's say what happens, not what doesn't happen

Very challenging! I feel for our users

barriebyron replied:
oh man, now I went and rewrote a bit more for clarity

AdityaSripal commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,186 @@ parent:
+Modules can also choose to write application-specific acknowledgments upon processing a packet. This acknowledgment can be done synchronously on `OnRecvPacket` if the module processes packets as soon as they are received from IBC module. Or acknowledgments can be done asynchronously if module processes packets at some later point after receiving the packet.
 
-Regardless, this acknowledgement data is opaque to IBC much like the packet `Data` and will be treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgemnet in such a way that the sender module can decode it correctly. This should be decided through version negotiation during the channel handshake.
+Regardless, this acknowledgment data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgment so that the sender module can decode it correctly. This should be decided through version negotiation during the channel handshake.
```
```suggestion
Regardless, this acknowledgment data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgment so that the sender module can decode it correctly. The encoding should be negotiated between the two parties during version negotiation in the channel handshake.
```

barriebyron replied:
yes, much better @AdityaSripal (except "should" let's use "must" instead)

AdityaSripal commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,186 @@ parent:
+The acknowledgment can encode whether the packet processing succeeded or failed, along with additional information that allows the sender module to take appropriate action.
 
-Once the acknowledgement has been written by the receiving chain, a relayer will relay the acknowledgement back to the original sender module which will then execute application-specific acknowledgment logic using the contents of the acknowledgement. This may involve rolling back packet-send changes in the case of a failed acknowledgement (refunding senders).
+After the acknowledgment has been written by the receiving chain, a relayer relays the acknowledgment back to the original sender module which  then executes application-specific acknowledgment logic using the contents of the acknowledgment. This can involve rolling back packet-send changes in the case of a failed acknowledgment (refunding senders).
```
```suggestion
After the acknowledgment has been written by the receiving chain, a relayer relays the acknowledgment back to the original sender module which  then executes application-specific acknowledgment logic using the contents of the acknowledgment. Application developers can roll back packet-send changes in the case of a failed acknowledgement (e.g. refunding senders in ICS20).
```

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,205 @@ parent:
 ### [Packets](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel)
 
-Modules communicate with each other by sending packets over IBC channels. As mentioned above, all
+Modules communicate with each other by sending packets over IBC channels. All
```
is there a reason we don't use the packet definition from https://github.com/cosmos/ibc/blob/master/spec/core/ics-004-channel-and-packet-semantics/README.md 

(seems much more clear!)

AdityaSripal replied:
I'm in favor of just copying that over.

The generic `Identifier` type in the spec should be changed to `string` for the go implementation

barriebyron replied:
@AdityaSripal can you take on that spec work?

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,205 @@ parent:
-must process a packet. If the timeout passes without the packet being successfully received, the
-sending module can timeout the packet and take appropriate actions.
+`channelID`. This packet structure allows modules to know the sender module of a given packet. IBC packets 
+contain a sequence to optionally enforce ordering. 
```
so the packet sequence that defines ordering is 
  `sequence: uint64` ?
  seems like we would want to define what is ORDERED and what is UNORDERED

AdityaSripal replied:
Yes if a channel is ORDERED, then the packets must be received in the order of the `sequence` that is inside the packet.

barriebyron commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,205 @@ parent:
+If the timeout is reached, then a proof of packet timeout can be submitted to the original chain. The original chain can then perform 
+application-specific logic to timeout the packet, perhaps by rolling back the packet send changes (refunding senders any locked funds, etc.).
+
+- In ORDERED channels, a timeout of a single packet in the channel causes the channel to close. 
```
since line 175 defines what happens for a packet timeout in an ORDERED channel, I attempted to describe what happens in an UNORDERED channel  on line 180

(we really gotta show this essential content some love... am I on the right track here?)

barriebyron commented:
hi @crodriguezvega thank you so much for your attention. This essential content can still be improved, but I think we're getting closer. 

I'm here tomorrow Friday Nov 19 (Eastern time) but then I'm off all next week. Feel free to make changes as you need and merge in at will. 

@AdityaSripal you've been awesome with you input, thank you! @damiannolan thank you for weighing in on word choice.

Please give the IBC content the attention and care it requires in this PR, I don't want to hold up iterative improvements! We can use the netlify preview and have someone read it through for a fresh look, if that helps! 

(there is still some repetition)

AdityaSripal commented on docs/ibc/overview.md:
```diff
@@ -6,170 +6,205 @@ parent:
+- The `timeoutHeight` indicates a consensus height on the destination chain after which the packet is no longer be processed, and instead counts as having timed-out.
+- The `timeoutTimestamp` indicates a timestamp on the destination chain after which the packet is no longer be processed, and instead counts as having timed-out.
+
+If the timeout passes without the packet being successfully received, the
+sending module can timeout the packet and take appropriate actions.
```
```suggestion
If the timeout passes without the packet being successfully received, the packet can no longer be
received on the destination chain. The sending module can timeout the packet and take appropriate actions.
```

AdityaSripal reviewed (approved):
Since @barriebyron is on vacation next week, I suggest that she make final edits to this PR based on any remaining suggestions, and then we can merge this PR.

@barriebyron , if you have any remaining questions/suggestions that you cannot resolve on your own, then please leave them as comments on this PR and we can address them in a subsequent PR/issue.

Thanks for the help!

barriebyron commented:
thanks everyone for the great collaboration! Please merge when tests complete!
